### PR TITLE
Split gated FC1 updates in Muon

### PR DESCRIPTION
**Pull request by mkhona-nvidia**
## What changed

- Detect 2D `linear_fc1` weights when `gated_linear_unit` is enabled and attach TP-local gate/up split metadata.
- Split fused GLU/SwiGLU FC1 updates into gate and up matrices before Muon orthogonalization, then restore the original fused layout.
- Apply the same behavior to TensorParallelAdaptiveMuon and preserve the metadata when Megatron copies tensor-parallel parameters.
- Add focused tests for split-shape validation, independent orthogonalization, optimizer-factory tagging, and metadata copying.

## Why

A gated FC1 is emitted as one larger GEMM, but its two halves are independent gate and up projections. Orthogonalizing the fused matrix as a single Muon update couples those projections and uses the wrong matrix geometry. Each half now receives its own Newton–Schulz orthogonalization and Muon scale.

## Impact

Gated MLPs, including SwiGLU, automatically use separate Muon updates for the gate and up projections. Non-gated models and non-FC1 parameters are unchanged.

## Validation

- `python3 -m compileall` on all modified Python files
- `git diff --check`
- Added targeted GPU/distributed unit coverage (not executed locally because this host does not have the repository's required CUDA container, `uv`, PyTorch, or pytest runtime)

### CODE REVIEW

File: megatron/core/transformer/moe/shared_experts.py [hunk 1]
```diff
@@ -132,6 +132,11 @@ def __init__(
             name=name,
             pg_collection=pg_collection,
         )
+        set_glu_linear_fc1_attributes(
+            self.linear_fc1,
+            self.config.gated_linear_unit,
+            self.config.moe_shared_expert_glu_interleave_size,
```
Comment (FDecaYed) [inline megatron/core/transformer/moe/shared_experts.py]: `SharedExpertMLP` always applies `moe_shared_expert_glu_interleave_size`, but that layout is only used when `use_grouped_gemm_for_shared_expert=True`. On the normal path FC1 remains contiguous, so Muon would incorrectly deinterleave the gradient. Could we only attach this metadata in `FusedSharedExpertMLP` and cover both paths in a test?

File: megatron/core/transformer/mlp.py [hunk 2]
```diff
@@ -84,6 +84,23 @@ def __call__(
...
+    """Describe the gate/up row layout on fused GLU FC1 weights."""
+    if not enabled:
+        return
+
+    for name, param in cast(torch.nn.Module, linear_fc1).named_parameters():
+        leaf_name = name.rsplit('.', maxsplit=1)[-1]
+        is_weight = leaf_name == 'weight' or (
+            leaf_name.startswith('weight') and leaf_name[len('weight') :].isdigit()
+        )
+        if is_weight and param.ndim == 2:
```
Comment (FDecaYed) [inline megatron/core/transformer/mlp.py:101]: With `moe_single_grouped_weight=True`, local expert weights are stored in one 3D `GroupedTensor`. The current 2D-only metadata check skips it, and `_is_nonlinear_or_embedding` then routes the weight to the scalar optimizer instead of Muon. Could we either support this layout or explicitly warn/reject this combination?

File: megatron/core/transformer/moe/shared_experts.py (hunk 1 above)
Comment (mkhona-nvidia) [reply megatron/core/transformer/moe/shared_experts.py]: Fixed in a3df7bbde by moving the interleave metadata override into FusedSharedExpertMLP. Plain SharedExpertMLP now retains the contiguous layout metadata inherited from MLP. I also added test_shared_expert_glu_metadata_matches_stored_layout, covering both paths.

File: megatron/core/transformer/mlp.py (hunk 2 above)
Comment (mkhona-nvidia) [reply megatron/core/transformer/mlp.py:101]: I chose to reject this combination explicitly rather than provide partial 3D support. Muon and Adaptive Muon now raise a clear ValueError when moe_single_grouped_weight=True, before the parameters can silently fall back to the scalar optimizer. Supporting it properly will require per-expert Muon state and orthogonalization within the 3D GroupedTensor. A regression test covers the rejection.

File: megatron/core/optimizer/__init__.py [hunk 3]
```diff
@@ -799,6 +809,36 @@ def _get_megatron_emerging_optimizer(
...
+                is_gtp_weight = getattr(param, 'is_gtp_weight_remat', False)
+                gtp_remat_size = getattr(param, 'gtp_remat_size', 1) if is_gtp_weight else 1
+                pad_length = getattr(param, 'pad_length', 0) if is_gtp_weight else 0
+                logical_rows = param.shape[0] * gtp_remat_size - pad_length
+                valid_interleave = interleave_size is None or interleave_size > 0
+                layout_multiple = 2 if interleave_size is None else 2 * interleave_size
+                if valid_interleave and logical_rows > 0 and logical_rows % layout_multiple == 0:
+                    param.is_glu = True
+                    param.glu_gtp_remat_size = gtp_remat_size
+                    param.glu_gtp_pad_length = pad_length
```
Comment (fanshiqing) [inline megatron/core/optimizer/__init__.py]: I don't think it's necessary to define new attributes here: `glu_gtp_remat_size` and `glu_gtp_pad_length`, just reusing the exisitng `gtp_reamt_size` and `pad_length` is enough.

Can you explain why we  have to add these two attributes?

File: megatron/core/optimizer/emerging_optimizers.py [hunk 4]
```diff
@@ -294,6 +297,120 @@ def scaled_orthogonalize_fn_with_gtp_remat(self, p, grad, tp_group, partition_di
         shard_size = gathered_grad.shape[0] // gtp_remat_size
         return gathered_grad[gtp_rank * shard_size : (gtp_rank + 1) * shard_size].contiguous()
 
+    def _gather_glu_grad(self, p, grad):
```
Comment (fanshiqing) [inline megatron/core/optimizer/emerging_optimizers.py:514]: this function is doing more branching than the logic needs.
let's make it simple:
- normalize `gtp_remat_size = get_pg_size(gtp_remat_group) if gtp_remat_group is not None else 1`, 
- validate once against `expected_size`, 
- then a single `if gtp_remat_size > 1: all_gather ... else: gathered_grad = grad`.

File: megatron/core/optimizer/emerging_optimizers.py [hunk 5]
```diff
@@ -294,6 +297,120 @@ def scaled_orthogonalize_fn_with_gtp_remat(self, p, grad, tp_group, partition_di
...
+            gtp_rank = get_pg_rank(gtp_remat_group)
+            if gtp_remat_size != expected_size:
+                raise RuntimeError(
+                    "Muon GLU split GTP size mismatch: "
+                    f"parameter metadata={expected_size}, process group={gtp_remat_size}"
+                )
+            if gtp_remat_size > 1:
+                shards = [torch.empty_like(grad) for _ in range(gtp_remat_size)]
+                torch.distributed.all_gather(shards, grad, gtp_remat_group)
+                gathered_grad = torch.cat(shards, dim=0)
```
Comment (fanshiqing) [inline megatron/core/optimizer/emerging_optimizers.py]: line 325-327 is just something dupliced with what's implemented in `scaled_orthogonalize_fn_with_gtp_remat`, 
let's do code reuse when possible

**Comment by FDecaYed** [comment]
`use_kitchen` still looks incorrect with TP > 1. `MLP` uses `fc1_stride=1` for Kitchen, so a local TP shard is a contiguous slice of the global `[gate; up]` rows rather than a local `[gate; up]` pair. However, `set_glu_linear_fc1_attributes` still tags it as `is_glu=True`, and the optimizer name fallback also ignores `partition_stride`. The current local half-split can therefore silently split gate-only or up-only rows as if they were separate gate/up matrices. Could we explicitly skip or reject the stride-1 layout until a gather-first implementation supports it?

**Review by FDecaYed — APPROVED** [review]
